File: 2016/Day8/main.lua
local Display = {}
Display.__index = Display

function Display:new()
    self = setmetatable({}, Display)
    self.WIDTH = 50
    self.HEIGHT = 6

    -- Initialize the screen with all pixels off
    self.screen = {}
    for y = 1, self.HEIGHT do
        self.screen[y] = {}
        for x = 1, self.WIDTH do
            self.screen[y][x] = false
        end
    end

    return self
end

---@param width number
---@param height number
function Display:rect(width, height)
    for y = 1, height do
        for x = 1, width do
            self.screen[y][x] = true
        end
    end
end

---@param row number
---@param amount number
function Display:rotate_row(row, amount)
    -- Save the original row
    local oldRow = {}
    for x = 1, self.WIDTH do
        oldRow[x] = self.screen[row][x]
    end

    -- Rotate the row
    for x = 1, self.WIDTH do
        self.screen[row][(x + amount - 1) % self.WIDTH + 1] = oldRow[x]
    end
end

---@param col number
---@param amount number
function Display:rotate_column(col, amount)
    -- Save the original column
    local oldCol = {}
    for y = 1, self.HEIGHT do
        oldCol[y] = self.screen[y][col]
    end

    -- Rotate the column
    for y = 1, self.HEIGHT do
        self.screen[(y + amount - 1) % self.HEIGHT + 1][col] = oldCol[y]
    end
end

---@return number
function Display:count_lit_pixels()
    local count = 0
    for y = 1, self.HEIGHT do
        for x = 1, self.WIDTH do
            if self.screen[y][x] then
                count = count + 1
            end
        end
    end
    return count
end

function Display:print_screen()
    for y = 1, self.HEIGHT do
        local row = ""
        for x = 1, self.WIDTH do
            row = row .. (self.screen[y][x] and "#" or ".")
        end
        print(row)
    end
end

local display = Display:new()

for line in io.lines('input.txt') do
    local parts = {}
    for part in line:gmatch('%S+') do
        table.insert(parts, part)
    end

    if parts[1] == 'rect' then
        local rawWidth, rawHeight = parts[2]:match('(%d+)x(%d+)')
        local width, height = tonumber(rawWidth), tonumber(rawHeight)
        assert(width and height, 'Invalid input')
        display:rect(width, height)
    elseif parts[1] == 'rotate' then
        if parts[2] == 'row' then
            local row = tonumber(parts[3]:match('y=(%d+)'))
            local amount = tonumber(parts[5])
            assert(row and amount, 'Invalid input')
            display:rotate_row(row + 1, amount) -- +1 because Lua arrays are 1-indexed
        elseif parts[2] == 'column' then
            local col = tonumber(parts[3]:match('x=(%d+)'))
            local amount = tonumber(parts[5])
            assert(col and amount, 'Invalid input')
            display:rotate_column(col + 1, amount) -- +1 because Lua arrays are 1-indexed
        end
    end
end

print('Number of lit pixels: ' .. display:count_lit_pixels())
print('Display:')
display:print_screen()

File: 2016/Day8/main.py
import sys


class Display:
    WIDTH = 50
    HEIGHT = 6

    def __init__(self):
        self.screen = [
            [False for _ in range(self.WIDTH)] for _ in range(self.HEIGHT)
        ]

    def rect(self, width: int, height: int):
        for r in range(height):
            for c in range(width):
                if 0 <= r < self.HEIGHT and 0 <= c < self.WIDTH:
                    self.screen[r][c] = True

    def rotate_row(self, row_idx: int, amount: int):
        if not (0 <= row_idx < self.HEIGHT):
            return

        row = self.screen[row_idx]
        effective_amount = amount % self.WIDTH
        self.screen[row_idx] = row[-effective_amount:] + row[:-effective_amount]

    def rotate_column(self, col_idx: int, amount: int):
        if not (0 <= col_idx < self.WIDTH):
            return

        column_values = [self.screen[r][col_idx] for r in range(self.HEIGHT)]

        effective_amount = amount % self.HEIGHT
        rotated_column = column_values[-effective_amount:] + column_values[:-effective_amount]

        for r in range(self.HEIGHT):
            self.screen[r][col_idx] = rotated_column[r]

    def count_lit_pixels(self) -> int:
        return sum(pixel for row in self.screen for pixel in row)

    def print_screen(self):
        for row in self.screen:
            print(''.join(['#' if pixel else '.' for pixel in row]))

if __name__ == '__main__':
    display = Display()

    for line in sys.stdin:
        parts = line.strip().split()
        command = parts[0]

        if command == 'rect':
            if len(parts) > 1:
                dimensions = parts[1].split('x')
                if len(dimensions) == 2:
                    width = int(dimensions[0])
                    height = int(dimensions[1])
                    display.rect(width, height)
        elif command == 'rotate':
            if len(parts) >= 5:
                op_type = parts[1]
                index_str = parts[2].split('=')[1]
                index = int(index_str)
                amount = int(parts[4])

                if op_type == 'row':
                    display.rotate_row(index, amount)
                elif op_type == 'column':
                    display.rotate_column(index, amount)

    print(f'Number of lit pixels: {display.count_lit_pixels()}')
    print('Final screen:')
    display.print_screen()
